[Observability Onboarding] Export package list from fleet

justinkambic opened the pull request:
## Summary

Resolves https://github.com/elastic/kibana/issues/178945.
Resolves #178947.
Resolves https://github.com/elastic/kibana/issues/178952.

~Undergoing the work to export package list in a sharable way while minimally impacting the existing implementation used by the Fleet team.~


### Update

Updated to provide the flows for all three paths of the form. Synthetics and APM still link to their respective apps. Logs icons link to the embedded flows.

#### Things to test

- Check that the integration cards behave in a manner you'd expect. There's a lot of custom card creation happening and it's possible I have made a mistake in some linking.
- Icons: there are a few icons that don't correspond to the design. I wasn't able to find the correct `src` for the colored-in APM logo, for example, and I'm not sure where the one logs icon lives. I've used the `logoFilebeat` icon in its place as it's quite similar and is heavily used in the integrations list.
- For the searchable integrations list, I was unsure if I should re-hide it when the user changes their radio selection in the top-level question. Right now it'll just stay as-is with their original query.
- The `Collections` feature, make sure collection buttons properly apply the search query.

~This now works with the exported package list grid from Fleet. I've introduced a real package list in place of the dummy grid that we had before. I also added a searchable grid below.~

~Surely there are some things to iron out still. I also still consider the code rough, so if you should choose to review it you will likely find areas you want to see improved.~

![20240405171834](https://github.com/elastic/kibana/assets/18429259/b3e4553a-1841-4c57-9713-81571fae1b44)

---

**Note:** I will likely extract the Fleet-specific pieces of this to a separate patch that we can merge to Kibana `main` to make it easier for the Fleet team to review the changes. We can then merge those upstream and pull them into our feature branch.

^^ this is still true, before we merged this I'll likely pull out the Fleet-specific modifications we need and have the Fleet team review them separately. We can still merge this to the feature branch in the meantime and resolve it later.

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

flash1293 commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/experimental_onboarding_flow.tsx:
```diff
@@ -4,17 +4,53 @@
+const queryClient = new QueryClient();
+
+function CustomPackageList() {
+  const availablePackages = useAvailablePackages({
```
Pretty nice this works as easily as that - not many dependencies to consider.

justinkambic replied:
I spoke with Kyle from Fleet and he's ++ on this approach. So, I think we can function as middleware here between that hook and the component. We will need some upstream modifications but they should be pretty unobtrusive.

justinkambic replied:
> Should we do that in any case and manage the search input in onboarding? I don't know whether we want to keep the "Search for integrations" placeholder in there. Of course we could also make this configurable, but the GridColumn + the useAvailablePackages hook seem like a pretty good level of abstraction for our use case.

++ I am thinking this is what we'll do. We can implement our own search logic (it will probably mimic what they have closely), or we can choose to enable their search bar and use it with the default behavior. I think we are already very close to the modifications we need upstream and can probably open a separate PR soon.

flash1293 replied:
Thanks, let me know when you are ready for a more thorough review.

flash1293 reviewed (commented):
Looks pretty good already - I like how little dependencies need to be wired up for this to work.

Some thoughts:
* Should we add a prop to hide the search bar, or maybe just export the GridColumn component as well for the top 6 display?
* Should we do that in any case and manage the search input in onboarding? I don't know whether we want to keep the "Search for integrations" placeholder in there. Of course we could also make this configurable, but the GridColumn + the useAvailablePackages hook seem like a pretty good level of abstraction for our use case.

justinkambic reviewed (commented):
@flash1293 Let me know what you think about the fleet-specific changes here. In essence this is a draft of what we would propose to them in an upstream PR. 

These changes allow us to override the cards' click handlers, specify an accent border + badge per the design, and exports the list and hook needed to make use of the component.

Once we have the upstream changes approved we can finish implementing the list's inclusion in our flow.

I have a quick demo usage using some overrides from the props like this:

<img width="559" alt="image" src="https://github.com/elastic/kibana/assets/18429259/d8061af2-0b17-4e60-86f6-50581cf50238">

![20240326170751](https://github.com/elastic/kibana/assets/18429259/04fb58b6-916e-4c8a-9dff-bd400d4a26e7)

flash1293 commented:
This looks great to me @justinkambic - to me this is ready for a look by the integrations team, wdyt?

One note - the endless scroll implementation doesn't really play well with our footered layout - it flickers in when scrolling:

![Kapture 2024-03-27 at 16 34 17](https://github.com/elastic/kibana/assets/1508364/49612b3f-b446-4d5f-9630-a79d1a427d52)

I guess when integrating it properly we need to give a container a max height to avoid this.

thomheymann commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/experimental_onboarding_flow.tsx:
```diff
@@ -13,32 +20,72 @@ import { Footer } from './footer/footer';
-    </>
+    <QueryClientProvider client={queryClient}>
+      <EuiThemeProvider darkMode={false}>
+        <>
```
I think you can remove the Fragment, unless EuiThemeProvider only allows a single child

justinkambic replied:
Sorry, the code on our side hasn't been finalized/is still in draft form so it's not production-worthy yet. I will work on this a bit hopefully today or tomorrow.

justinkambic replied:
7e699357bafb9287bbda7c1347a893c6ed92abf9

thomheymann commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/experimental_onboarding_flow.tsx:
```diff
@@ -13,32 +20,72 @@ import { Footer } from './footer/footer';
+            color="subdued"
+            restrictWidth
+          >
+            <OnboardingFlowForm />
+            <EuiSpacer size="xl" />
+            <OnboardingFlowPackageList />
```
Is there a reason to render the package list outside of the form? In the design it's in the same panel and we need to change the filter based on the form selection so it might be easier to move it into the form. 

The spacing below the form was intentional to separate the content from the footer correctly so I would keep that in.

thomheymann commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/experimental_onboarding_flow.tsx:
```diff
@@ -13,32 +20,72 @@ import { Footer } from './footer/footer';
-      </EuiPageTemplate.Section>
-    </>
+    <QueryClientProvider client={queryClient}>
+      <EuiThemeProvider darkMode={false}>
```
The observability app is already rendering a theme provider? 

Is there a reason you are rendering another one inside the experimental onboarding flow component and hard coding the theme to light mode? This will break Kibana and user settings.

justinkambic replied:
Sorry, the code on our side hasn't been finalized/is still in draft form so it's not production-worthy yet. I will work on this a bit hopefully today or tomorrow. ++ we definitely don't want to do this, I just wrote that quickly to resolve errors when initially importing the package list.

justinkambic replied:
7e699357bafb9287bbda7c1347a893c6ed92abf9

thomheymann reviewed (commented):
Looks great! Couple comments below.

thomheymann reviewed (commented):
Looks great but there's a few issues with the async states:

1. There are no loading indicators. We should show a loading spinner while the async bundle and integrations are loaded in.
2. Async operations can fail and there are currently no error states. We need to show an error message and ideally a retry button when the async chunk or integrations can't be loaded. 
3. Custom cards are already displayed while integrations are still loading. Once integrations have loaded all cards in the grid are then re-sorted causing UI jankyness. We should either leave custom cards in the same position or wait until the other integrations have been loaded before showing them if we want them sorted together with the integration cards. 
4. When clicking through to an integration the "< Back to integrations" link takes me to the integrations page. It would be good to be able to either hide this or change this button so it takes the user back to the onboarding flow. (Can be a follow up issue)

https://github.com/elastic/kibana/assets/190132/88fca5a8-ee40-4778-bf59-cd2188506e31

justinkambic commented:
@thomheymann I agree about the loading states for the integration list. I have updated this so that we will display a loading state if the packages are loading _and_ the initial hide feature isn't in place. This way we are displaying the search box only in the bottom grid. If they were to input a query before the packages loaded, they'd see the loading indicator as well. And in cases where we _do_ display the grid, we'll now skip rendering until the full list is built so it won't be janky. Here's an example of the loading state UI I added, and a practical run-through:

## Loading state UI
![20240409130526](https://github.com/elastic/kibana/assets/18429259/772d63c4-256d-4130-af0b-3c243cb22a9b)

## Practical loading UX
![20240409130840](https://github.com/elastic/kibana/assets/18429259/491dd0af-6b31-48da-9830-a9d1491dd737)

I noticed in the design we have the skeleton rectangle loading indicators. I had some trouble getting these to adhere to flex principles, and I didn't want to waste more time playing with it.

> We should show a loading spinner while the async bundle and integrations are loaded in.

I disagree on the loading spinner for the async bundle. Practically, I'd expect this to be pretty fast and it will make things look janky again if we go from spinner to the loading elements displayed later while fetching integrations. I did experiment with re-using the same loading component in both places and it seems to be pretty smooth for the curated list. In the case of the list where we show the search bar (and therefore no loading state, unless they start to query before data is fetched) it again makes for a janky experience. I'd propose we either conditionally show the loading state only when the search bar isn't the default element, or just skip expressing this loading, as it should be quick and in the case where it fails we will show a retry flow.

> We need to show an error message and ideally a retry button when the async chunk or integrations can't be loaded.

Here's the experience I put together for this:

![20240409160742](https://github.com/elastic/kibana/assets/18429259/adbfeec4-4b1f-4701-95d7-112bfcb5255d)

> Can be a follow up issue

We should make a follow-up issue for this. It'll require upstream modifications again and is best done outside of our feature branch. Then we can consume the new feature on our side.

flash1293 commented:
Looks pretty good overall 🎉 

About the navigation question: I agree that we can do the "Back to navigation" thing in the integrations plugin as a followup, the same thing is also a problem when using the browser back button because the choice of the first question is not persisted which is something I expected.

I made the same comment on @thomheymann s PR, so make sure to only resolve it once :) I think persisting this choice in the path is probably the easiest way to have consistent behavior.

No strong opinion on the loading state for fetching the bundles - this is not really handled in a consistent way today throughout the application.

> I disagree on the loading spinner for the async bundle. Practically, I'd expect this to be pretty fast and it will make things look janky again if we go from spinner to the loading elements displayed later while fetching integrations

React suspense will hold back the loading state for a little bit to deal with this - IMHO it's not a blocker though

justinkambic commented:
> React suspense will hold back the loading state for a little bit to deal with this - IMHO it's not a blocker though

I wasn't specific enough in my response there; we have to also `async` load the hook function that pulls down the integrations from the API and this is where I don't necessarily think we want to force a loading state. I _suspect_ this will be quite fast in most cases and we are now handling the error condition, so the happy path will be mostly transparent to the user.

For the integrations list we are supplying Suspense with the placeholder so I am less concerned about that case.

justinkambic commented:
@flash1293 @thomheymann I have updated this branch with the collections functionality. I will again merge the Fleet changes separately upstream but the behavior seems to be consistent with what we discussed earlier in the call.

![20240410132030](https://github.com/elastic/kibana/assets/18429259/e0449bfd-aed1-48bb-b4c3-cae5b61ad959)

flash1293 commented:
Looks great @justinkambic ! @thomheymann should we merge this PR first and then you rebase yours or the other way around?

thomheymann commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/packages_list/index.tsx:
```diff
@@ -0,0 +1,201 @@
+    const [errorLoading, setErrorLoading] = useState(false);
+    const [loadingModule, setLoadingModule] = React.useState(true);
+
+    useEffect(() => {
+      async function load() {
+        setErrorLoading(false);
+        setLoadingModule(true);
+        try {
+          ref.current = await fetchAvailablePackagesHook();
+        } catch (e) {
+          setErrorLoading(true);
+        } finally {
+          setLoadingModule(false);
+        }
+      }
+      load();
+    }, [reloadRetry]);
```
I would suggest using a hook like `useAsyncRetry` from `react-use` package for loading and error states which handles all the complexity around it for you: https://streamich.github.io/react-use/?path=/story/side-effects-useasyncretry--docs

Trying to implement this manually often causes issues for weird edge cases. For example an async function could resolve when React has already unmounted the component causing an error. Your manual logic here would not handle this scenario correctly.

justinkambic replied:
TIL about `useAsyncRetry` - I really need to sit down and read all the docs for `react-use` sometime. So many useful shortcuts.

I will change this ++

justinkambic replied:
So much cleaner https://github.com/elastic/kibana/pull/179301/commits/c2dfb88cb92cbfb471aa316f515610ce473f4449

thomheymann reviewed (commented):
Looks great! Left one comment regarding the async hook implementation.

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

flash1293 commented:
Looks almost perfect, thanks @justinkambic ! (didn't look too deep into the code)

In terms of functionality:
* Nginx is not a quickstart yet, let's remove this label for now
* Could we remove the red buttons on top to jump to the system logs / custom logs flow (as this is captured by the regular UI now)

flash1293 commented:
/oblt-deploy-serverless

flash1293 commented:
Let's try to get this one merged quickly so we have a version of the new landingpage we can show outside of our team

thomheymann reviewed (approved):
Thanks for this, looking good so far. 

There's a few bugs and UX concerns but in the interest of being able to parallelise work let's get this merged in. 

Outstanding issues / work as far as I can see:
- bug: Grid showing incorrect Quickstart flows (only system logs and custom logs should be shown as quickstart flow)
- bug: After entering a search term manually, the search field does not show the search term correctly when selecting a collection
- bug: Selecting a collection should scroll to and focus the grid, not the search term field (scrolling should be smooth)
- UX: clearing search term shows all integrations instead of clearing the grid?
- UX: Quickstart flows (e.g. system logs) are not searchable but integrations are (e.g. nginx)
- Search term should be highlighted in search results
- Remove temporary logs selection header
- Selecting a user case should scroll smoothly to the grid
- Add ability to return from integrations page to onboarding page

Again, let's discuss / tackle these separately.
